vizro-core/src/vizro/figures/kpi_cards.py

```diff
@@ -0,0 +1,167 @@
+"""Contains default KPI card functions."""
+
+import warnings
+from typing import Optional
+
+import dash_bootstrap_components as dbc
+import numpy as np
+import pandas as pd
+from dash import html
+
+from vizro.models.types import capture
+
+FORMATTING_WARNING = (
+    "Custom format string detected. Note that the format string is being evaluated, so ensure that "
+    "only trusted user input is provided to prevent potential security risks."
+)
+
+VALUE_FORMAT_DEFAULT = "{value}"
+REFERENCE_FORMAT_DEFAULT = "{delta_relative:.1%} vs. reference ({reference})"
+
+
+@capture("figure")
+def kpi_card(  # noqa: PLR0913
+    data_frame: pd.DataFrame,
+    value_column: str,
+    *,
+    value_format: str = VALUE_FORMAT_DEFAULT,
+    agg_func: str = "sum",
+    title: Optional[str] = None,
+    icon: Optional[str] = None,
+) -> dbc.Card:
+    """Creates a styled KPI (Key Performance Indicator) card displaying a value.
+
+    Args:
+        data_frame: Data frame containing the data.
+        value_column: Column name of the value to be shown.
+        value_format: Format string to be applied to the value. It must be a valid Python format string where any of the
+            below placeholders can be used. Defaults to "{value}".
+            - value: `value_column` aggregated by `agg_func`.
+
+            For more details see: https://docs.python.org/3/library/string.html#format-specification-mini-language
+
+
+            Common examples include:
+             - "{value}": Displays the raw value.
+             - "${value:0.2f}": Formats the value as a currency with two decimal places.
+             - "{value:.0%}": Formats the value as a percentage without decimal places.
+             - "{value:,}": Formats the value with comma as a thousands separator.
+
+        agg_func: String function name to be used for aggregating the data. Common options include
+            "sum", "min", "max", "mean" or "median". Default is "sum".
+        title: KPI title displayed on top of the card. If not provided, it defaults to the capitalized `value_column`.
+        icon: Name of the icon from the [Google Material Icon Library](https://fonts.google.com/icons) to be displayed
+            on the left side of the KPI title. If not provided, no icon is displayed.
+
+    Raises:
+        UserWarning: If `value_format` is provided, a warning is raised to make aware that only trusted user
+            input should be provided.
+
+    Returns:
+         A Dash Bootstrap Components card (`dbc.Card`) containing the formatted KPI value.
+
+    """
+    if value_format != VALUE_FORMAT_DEFAULT:
```

Comment:
Just to be clear, the "untrusted input" here could be for two reasons:
1. you configure your dashboard using yaml and the _dashboard developer_ who has access to writing the yaml configuration is not as trusted as the developer who writes the Python (similar logic to kedro config)
2. you accept input from the _dashboard user_, e.g. by letting `value_format` be a `vm.Parameter` (same risk as at bottom of page [here](https://vizro.readthedocs.io/en/stable/pages/user-guides/custom-actions/))

The reason the case here is worth pointing out in particular is because it's not at all obvious that someone able to specify what looks like a pure and innocent formatting string could potentially write something that compromises security. (unlike `eval` for example where it's kind of obvious that it's insecure to accept any string input)

---

However, both the above are really edge cases and on second thoughts I think raising a `UserWarning` is probably going to be too alarmist/cautious and raise more questions than it solves. I like that you excluded the default case (otherwise it would be super annoying) but think this will still be a bit too annoying. The way you've written it that introduces constant variables makes sense but also makes it more difficult for people to copy and paste the code to write their own KPI functions, which I think is also an important consideration here.

So what I'd suggest is that we instead just put the warning in the API docs on these two functions, linking to https://stackoverflow.com/questions/76783239/is-it-safe-to-use-python-str-format-method-with-user-submitted-templates-in-serv. No need to explain all the above stuff about what is meant by "untrusted user input", that was just to explain where I'm coming from in case it wasn't clear before.